src/coreclr/vm/codeman.cpp

```diff
@@ -4605,118 +4621,140 @@ RangeSection* ExecutionManager::GetRangeSection(TADDR addr)
-        pHead->pLastUsed = pLast;
-#endif
+        rh->last_used_index = LastUsedRSIndex;
+#endif //DACCESS_COMPILE
 
-    return pCurr;
+    return (foundIndex>=0)?rh->array[foundIndex].pRS:NULL;
 }
 
-RangeSection* ExecutionManager::GetRangeSectionAndPrev(RangeSection *pHead, TADDR addr, RangeSection** ppPrev)
+/*********************************************************************/
+// This function returns coded index: a number to be converted 
+// to an actual index of RangeSectionHandleArray. Coded index uses
+// non-negative values to store actual index of an element already
+// existed in the array or it uses negative values to code an index
+// to be used for placing a new element to the array. If value
+// returned from FindRangeSectionHandleHelper is not negative, it can
+// be immediately used to read an existing element like this:
+// pRS = RangeSectionHandleArray[non_negative].pRS;
+// If the value becomes negative then there is no element with
+// requested TADDR in the RangeSectionHandleArray, and the value
+// might be decoded to an actual index of RangeSectionHandleArray cell
+// to place new RangeSectionHandle (if intended).
+// The function to decode is DecodeRangeSectionIndex that takes
+// negative coded index and returns an actual index.
+//
+int ExecutionManager::FindRangeSectionHandleHelper(RangeSectionHandleHeader *h, TADDR addr)
 {
-    WRAPPER_NO_CONTRACT;
+    CONTRACTL {
+        NOTHROW;
+        HOST_NOCALLS;
+        GC_NOTRIGGER;
+        SUPPORTS_DAC;
+    } CONTRACTL_END;
+
+    RangeSectionHandle *array = h->array;
 
-    RangeSection *pCurr;
-    RangeSection *pPrev;
-    RangeSection *result = NULL;
+    _ASSERTE(array != nullptr);
+    _ASSERTE(h->capacity != (SIZE_T)0);
 
-    for (pPrev = NULL,  pCurr = pHead;
-         pCurr != NULL;
-         pPrev = pCurr, pCurr = pCurr->pnext)
+    if ((int)h->size == 0)
     {
-        // See if addr is in [pCurr->LowAddress .. pCurr->HighAddress)
-        if (pCurr->LowAddress > addr)
-            continue;
+        return EncodeRangeSectionIndex(0);
+    }
 
-        if (addr >= pCurr->HighAddress)
-            break;
+    int iLow = 0;
+    int iHigh = (int)h->size - 1;
+    int iMid = (iHigh + iLow)/2;
 
-        // addr must be in [pCurr->LowAddress .. pCurr->HighAddress)
-        _ASSERTE((pCurr->LowAddress <= addr) && (addr < pCurr->HighAddress));
+    if(addr < array[0].LowAddress)
+    {
+        return EncodeRangeSectionIndex(0);
+    }
 
-        // Found the matching RangeSection
-        result = pCurr;
+    if (addr >= array[iHigh].pRS->HighAddress)
+    {
+        return EncodeRangeSectionIndex(iHigh + 1);
+    }
 
-        // Write back pPrev to ppPrev if it is non-null
-        if (ppPrev != NULL)
-            *ppPrev = pPrev;
+    do
+    {
+        if (addr < array[iMid].LowAddress)
+        {
+            iHigh = iMid;
+            iMid = (iHigh + iLow)/2;
+        }
+        else if (addr >= array[iMid+1].LowAddress)
```

Comment:
The code here is doing a binary search of the RangeSection list, but I don't see how its protected from a separate thread writing to the sorted list. In particular, if there is an Add going on, which may be performing a memcpy of the list to insert something in the middle, the structure of the reader lock as written will allow a reader to enter, and read while the list is in an inconsistent state.

Reply:
Ah, I missed that. That highlights the concern I have with the comments, etc. ALL of those need to be updated to accurately describe the new model. Also, recent experience has shown that its much safer to use explicit volatile load/store methods instead, so the particular memory operations can easily be seen and examined.